cmd/pebble/ycsb.go

```diff
@@ -7,190 +7,477 @@ package main
-			if i%20 == 0 {
-				fmt.Println("optype__elapsed____ops/sec__p50(ms)__p95(ms)__p99(ms)_pMax(ms)")
-			}
-			reg.Tick(func(tick histogramTick) {
-				h := tick.Hist
-				fmt.Printf("%6s %8s %10.1f %8.1f %8.1f %8.1f %8.1f\n",
-					tick.Name,
-					time.Duration(elapsed.Seconds()+0.5)*time.Second,
-					float64(h.TotalCount())/tick.Elapsed.Seconds(),
-					time.Duration(h.ValueAtQuantile(50)).Seconds()*1000,
-					time.Duration(h.ValueAtQuantile(95)).Seconds()*1000,
-					time.Duration(h.ValueAtQuantile(99)).Seconds()*1000,
-					time.Duration(h.ValueAtQuantile(100)).Seconds()*1000,
-				)
-			})
-		},
-		done: func(elapsed time.Duration) {
-			fmt.Println("\noptype__elapsed_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)")
-			reg.Tick(func(tick histogramTick) {
-				h := tick.Cumulative
-				fmt.Printf("%6s %7.1fs %14d %14.1f %8.1f %8.1f %8.1f %8.1f %8.1f\n",
-					tick.Name, elapsed.Seconds(), h.TotalCount(),
-					float64(h.TotalCount())/elapsed.Seconds(),
-					time.Duration(h.Mean()).Seconds()*1000,
-					time.Duration(h.ValueAtQuantile(50)).Seconds()*1000,
-					time.Duration(h.ValueAtQuantile(95)).Seconds()*1000,
-					time.Duration(h.ValueAtQuantile(99)).Seconds()*1000,
-					time.Duration(h.ValueAtQuantile(100)).Seconds()*1000)
-			})
-
-			if empty := atomic.LoadInt64(&numEmptyResults); empty != 0 {
-				fmt.Printf("Number of reads that didn't return any results: %d.\n", empty)
+				b = db.NewBatch()
 			}
-			seq := atomic.LoadInt64(&seq.val)
-			var ch string
-			if ycsbConfig.sequential {
-				ch = "S"
-			} else {
-				ch = "R"
+			_ = b.Set(y.makeKey(uint64(i)), y.randBytes(rng), nil)
+		}
+		if err := b.Commit(y.writeOpts); err != nil {
+			log.Fatal(err)
+		}
+
+		y.keyNum = ackseq.New(uint64(ycsbConfig.initialKeys))
+	}
+
+	wg.Add(concurrency)
+	for i := 0; i < concurrency; i++ {
+		go y.run(db, wg)
+	}
+}
+
+func (y *ycsb) run(db *pebble.DB, wg *sync.WaitGroup) {
+	defer wg.Done()
+
+	rng := randvar.NewRand()
+	for {
+		start := time.Now()
+
+		op := y.ops.Int()
+		switch op {
+		case ycsbInsert:
+			y.insert(db, rng)
+		case ycsbRead:
+			y.read(db, rng)
+		case ycsbScan:
+			y.scan(db, rng)
+		case ycsbUpdate:
+			y.update(db, rng)
+		default:
+			panic("not reached")
+		}
+
+		y.latency[op].Record(time.Since(start))
+		if ycsbConfig.numOps > 0 &&
+			atomic.AddUint64(&y.numOps, 1) >= ycsbConfig.numOps {
+			break
+		}
+	}
+}
+
+func (y *ycsb) hashKey(key uint64) uint64 {
+	// Inlined version of fnv.New64 + Write.
+	const offset64 = 14695981039346656037
+	const prime64 = 1099511628211
+
+	h := uint64(offset64)
+	for i := 0; i < 8; i++ {
+		h *= prime64
+		h ^= uint64(key & 0xff)
+		key >>= 8
+	}
+	return h
+}
+
+func (y *ycsb) makeKey(keyNum uint64) []byte {
+	key := make([]byte, 4, 24)
+	copy(key, "user")
+	return strconv.AppendUint(key, y.hashKey(keyNum), 10)
+}
+
+func (y *ycsb) nextReadKey() []byte {
+	keyNum := y.hashKey(y.keyDist.Uint64()) % y.keyNum.Base()
```

Comment:
I've been trying to understand how we prevent the distribution from changing completely every time we get a new write. Or do we prevent that? A different value for `y.keyNum.Base()` looks like it'd change which keys are hot/cold.

Reply:
That does seem suspicious. This is the same logic in `workload/ycsb`. Let me see what the official YCSB implementation does.

Reply:
From `CoreWorkload.java`:

      // it does this by generating a random "next key" in part by taking the modulus over the
      // number of keys.
      // If the number of keys changes, this would shift the modulus, and we don't want that to
      // change which keys are popular so we'll actually construct the scrambled zipfian generator
      // with a keyspace that is larger than exists at the beginning of the test. that is, we'll predict
      // the number of inserts, and tell the scrambled zipfian generator the number of existing keys
      // plus the number of predicted keys as the total keyspace. then, if the generator picks a key
      // that hasn't been inserted yet, will just ignore it and pick another key. this way, the size of
      // the keyspace doesn't change from the perspective of the scrambled zipfian generator

Reply:
Which is to say that I think the code here is doing it wrong.

Reply:
Got it. Yeah, if we know in advance the size of the keyspace we can probably afford to regenerate, except the first few reads would require a lot of regenerations. We could also make the base used for reads the next power of two so reads require generating a random number at most twice (in expectation). Then the distribution changes every time row count surpasses the next power of two which might be infrequent enough.

Reply:
I think it is sufficient to simply get rid of the `y.hashKey` here. We're already hashing in `y.makeKey` so that hot items (small key nums) will be spread across the key space. Am I missing anything here? Is there a reason to hash before and after the mod.

Reply:
I don't know. Maybe it skews the distribution left (very slightly) by having the probability of `keyNum == 0` to be  `P[y.keyDist.Uint64() == 0] + P[y.keyDist.Uint64() == y.keyNum.Base()] + ...`, whereas `P[keyNum ==  y.keyNum.Base() - 1] = P[y.keyDist.Uint64() == y.keyNum.Base()-1] + P[y.keyDist.Uint64() == 2*y.keyNum.Base()-1]+...`.

Reply:
Because `P[y.keyDist.Uint64() == y.keyNum.Base()] > P[y.keyDist.Uint64() == 2*y.keyNum.Base()-1]` and there's no real reason to give the larger one to `P[keyNum == 0]`.

Reply:
cc @jeffrey-xiao

Reply:
I think I finally understand what they mean with that comment and what they described in section 5.3 of the paper. The SkewedZipfianGenerator has a keyspace of N + T * I + e, where N is the number of initial records, T is the number of operations, and I is the proportion of inserts. However, the SkewedZipfianGenerator hashes the value from a ZipfianGenerator with a very large key space (`10000000000` in our case). We should generate values using the SkewedZipfianGenerator until it yields a key that is known to exist (that can be determined using AcknowledgedCounter). This way, the keyspace of the SkewedZipfianGenerator never changes and therefore the distribution never changes as well. I'm not entirely sure why they hash it twice, but I'm guessing that they don't always want the hotspot to be the first inserted key.

So in practice, we shouldn't use IncrementIMax when the request distribution is zipfian (or have it do nothing since it's useful for the uniform generator).

Reply:
@jeffrey-xiao I think you meant `s/SkewedZipfGenerator/ScrambledZipfGenerator/g`.

> I'm not entirely sure why they hash it twice, but I'm guessing that they don't always want the hotspot to be the first inserted key.

I'm confused about the purpose of hashing before mod'ing and then hashing again after mod'ing. The zipf generator returns values in the range `[min,max)` where `min` is much more likely. For our configuration, `min == 1`. The `keyNum` is passed to `makeKey` where it is hashed which ensures that the hot keynums (values 1, 2, 3, 4, etc) are dispersed in the key space. So what is the purpose of hashing before the mod? 

@ajkr Yes, simply removing the hash before the mod would skew the key distribution slightly, but only ever so slightly. The first 10000 values (i.e. the default value of `--initial-keys`) returned by our zipf distribution capture 96% of the values. Increasing `keyNum.Base()` will not change the values for those hot values.

Reply:
Sorry, I realized I wasn't very clear with my point.

The workers will attempt to insert keys in a particular order. The first key attempted to be inserted is always the key with `keynum=1`, the second key is always the key with `keynum=2`, and so on. So the key with the highest probability of being read is always the first key attempted to be inserted. If the keynums are hashed before modding, then the probability of read keys are "randomized" w.r.t the insertion order. I'm not sure how big of an impact this would be, but this is what I initially thought the double hash was there for.